Meta/Lagom/Tools/CodeGenerators/LibPublicSuffix/GeneratePublicSuffixData.cpp

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) 2023, Cameron Youell <cameronyouell@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "../LibUnicode/GeneratorUtil.h"
+#include <AK/SourceGenerator.h>
+#include <AK/StringBuilder.h>
+#include <LibCore/ArgsParser.h>
+#include <LibCore/File.h>
+#include <LibMain/Main.h>
+
+ErrorOr<void> generate_header_file(Core::InputBufferedFile&, Core::File&);
+ErrorOr<void> generate_implementation_file(Core::InputBufferedFile&, Core::File&);
+
+ErrorOr<int> serenity_main(Main::Arguments arguments)
+{
+    StringView generated_header_path;
+    StringView generated_implementation_path;
+    StringView public_suffix_list_path;
+
+    Core::ArgsParser args_parser;
+    args_parser.add_option(generated_header_path, "Path to the header file to generate", "generated-header-path", 'h', "generated-header-path");
+    args_parser.add_option(generated_implementation_path, "Path to the implementation file to generate", "generated-implementation-path", 'c', "generated-implementation-path");
+    args_parser.add_option(public_suffix_list_path, "Path to the public suffix list", "public-suffix-list-path", 'p', "public-suffix-list-path");
+    args_parser.parse(arguments);
+
+    auto identifier_data = TRY(open_file(public_suffix_list_path, Core::File::OpenMode::Read));
+
+    auto generated_header_file = TRY(Core::File::open(generated_header_path, Core::File::OpenMode::Write));
+    auto generated_implementation_file = TRY(Core::File::open(generated_implementation_path, Core::File::OpenMode::Write));
+
+    TRY(generate_header_file(*identifier_data, *generated_header_file));
+    TRY(generate_implementation_file(*identifier_data, *generated_implementation_file));
+
+    return 0;
+}
+
+ErrorOr<void> generate_header_file(Core::InputBufferedFile&, Core::File& file)
+{
+    StringBuilder builder;
+    SourceGenerator generator { builder };
+    generator.append(R"~~~(
+#pragma once
+
+#include <AK/DeprecatedString.h>
+#include <AK/Forward.h>
+#include <AK/Trie.h>
+
+namespace PublicSuffix {
+
+class PublicSuffixData {
+protected:
+    PublicSuffixData();
+
+public:
+    PublicSuffixData(PublicSuffixData const&) = delete;
+    PublicSuffixData& operator=(PublicSuffixData const&) = delete;
+
+    static PublicSuffixData* the()
+    {
+        static PublicSuffixData* s_the;
+        if (!s_the)
+            s_the = new PublicSuffixData;
+        return s_the;
+    }
+
+    ErrorOr<Optional<String>> get_public_suffix(StringView string);
+
+private:
+    Trie<char, DeprecatedString> m_dictionary;
```

Comment:
Storing these as DeprecatedStrings in new code is a little gross. How about storing Strings, constructing the PublicSuffixData in a fallible factory function, and then making `the()` return ErrorOr?